File: lib/services/weather_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:intl/intl.dart';

class InspectionWindow {
  final DateTime startTime;
  final DateTime endTime;
  final double score;
  final double tempF;
  final double windMph;
  final double cloudCover;
  final double precipProb;
  final double humidity;
  final String condition; // Summary condition
  final List<String> issues; // Kill switch reasons or warnings
  final Map<String, int> scoreBreakdown; // Factor -> Points

  InspectionWindow({
    required this.startTime,
    required this.endTime,
    required this.score,
    required this.tempF,
    required this.windMph,
    required this.cloudCover,
    required this.precipProb,
    required this.humidity,
    required this.condition,
    required this.issues,
    required this.scoreBreakdown,
  });
}

class WeatherService {
  static const String _zipApiUrl = 'https://api.zippopotam.us/us/';
  static const String _weatherApiUrl = 'https://api.open-meteo.com/v1/forecast';

  Future<Map<String, double>> getCoordinates(String zip) async {
    final response = await http.get(Uri.parse('$_zipApiUrl$zip'));
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      final lat = double.parse(data['places'][0]['latitude']);
      final lng = double.parse(data['places'][0]['longitude']);
      return {'lat': lat, 'lng': lng};
    } else {
      throw Exception('Invalid ZIP code');
    }
  }

  Future<Map<String, dynamic>> getWeatherForecast(double lat, double lng) async {
    final uri = Uri.parse(
        '$_weatherApiUrl?latitude=$lat&longitude=$lng&hourly=temperature_2m,relative_humidity_2m,precipitation_probability,precipitation,weathercode,cloudcover,windspeed_10m&temperature_unit=fahrenheit&wind_speed_unit=mph&precipitation_unit=inch&timezone=auto&forecast_days=14');
    final response = await http.get(uri);
    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load weather data');
    }
  }

  List<InspectionWindow> calculateForecast(Map<String, dynamic> weatherData) {
    List<InspectionWindow> windows = [];
    final hourly = weatherData['hourly'];
    final times = hourly['time'] as List;
    final temps = hourly['temperature_2m'] as List;
    final humidities = hourly['relative_humidity_2m'] as List;
    final precipProbs = hourly['precipitation_probability'] as List;
    final precips = hourly['precipitation'] as List;
    final codes = hourly['weathercode'] as List;
    final clouds = hourly['cloudcover'] as List;
    final winds = hourly['windspeed_10m'] as List;

    // Group indices by Date (yyyy-MM-dd)
    Map<String, List<int>> dayIndices = {};
    for (int i = 0; i < times.length; i++) {
        DateTime t = DateTime.parse(times[i]);
        String dayKey = DateFormat('yyyy-MM-dd').format(t);
        if (!dayIndices.containsKey(dayKey)) dayIndices[dayKey] = [];
        dayIndices[dayKey]!.add(i);
    }

    final targetStartHours = [6, 8, 10, 12, 14, 16];

    // Iterate over each day
    for (var dayKey in dayIndices.keys) {
        List<int> indices = dayIndices[dayKey]!;
        
        for (int startHour in targetStartHours) {
            // Find index for this specific hour
            int? startIndex;
            for (int idx in indices) {
                if (DateTime.parse(times[idx]).hour == startHour) {
                    startIndex = idx;
                    break;
                }
            }

            // We need 2 hours of data: startHour and startHour + 1
            if (startIndex != null && (startIndex + 1) < times.length) {
                // Ensure the next hour is actually the next hour (data should be sequential, but good to be safe)
                // Assuming sequential for now as per OpenMeteo spec.
                
                int i = startIndex; 
                // Using 2 data points for 2-hour window
                List<double> segmentTemps = [temps[i], temps[i+1]].map((e) => (e as num).toDouble()).toList();
                List<double> segmentWinds = [winds[i], winds[i+1]].map((e) => (e as num).toDouble()).toList();
                List<double> segmentClouds = [clouds[i], clouds[i+1]].map((e) => (e as num).toDouble()).toList();
                List<double> segmentPrecipProbs = [precipProbs[i], precipProbs[i+1]].map((e) => (e as num).toDouble()).toList();
                List<double> segmentPrecips = [precips[i], precips[i+1]].map((e) => (e as num).toDouble()).toList();
                List<int> segmentCodes = [codes[i], codes[i+1]].map((e) => (e as num).toInt()).toList();
                List<double> segmentHumidities = [humidities[i], humidities[i+1]].map((e) => (e as num).toDouble()).toList();

                // Averages
                double avgTemp = segmentTemps.reduce((a, b) => a + b) / 2;
                double avgWind = segmentWinds.reduce((a, b) => a + b) / 2;
                double avgCloud = segmentClouds.reduce((a, b) => a + b) / 2;
                double avgPrecipProb = segmentPrecipProbs.reduce((a, b) => a + b) / 2;
                double avgHumidity = segmentHumidities.reduce((a, b) => a + b) / 2;

                // Kill Checks (Min/Max)
                double minTemp = segmentTemps.reduce((curr, next) => curr < next ? curr : next);
                double maxWind = segmentWinds.reduce((curr, next) => curr > next ? curr : next);
                double maxPrecipProb = segmentPrecipProbs.reduce((curr, next) => curr > next ? curr : next);
                double maxPrecipRate = segmentPrecips.reduce((curr, next) => curr > next ? curr : next);
                bool hasStorm = segmentCodes.any((c) => [95, 96, 99].contains(c));

                List<String> issues = [];
                if (minTemp < 55) {
                  issues.add("Too Cold (< 55°F)");
                }
                if (maxWind > 24) {
                  issues.add("Too Windy (> 24mph)");
                }
                if (maxPrecipProb > 49) {
                  issues.add("Rain Likely (> 49%)");
                }
                if (maxPrecipRate > 0.02) {
                  issues.add("Raining");
                }
                if (hasStorm) {
                  issues.add("Stormy Weather");
                }

                double totalScore = 0;
                Map<String, int> breakdown = {};

                // Scoring Logic (Always calculate breakdown)
                // Temp (Max 40)
                int tempScore = 0;
                if (avgTemp >= 75) {
                  tempScore = 40;
                } else if (avgTemp >= 70) {
                  tempScore = 37;
                } else if (avgTemp >= 65) {
                  tempScore = 33;
                } else if (avgTemp >= 60) {
                  tempScore = 27;
                } else if (avgTemp >= 57) {
                  tempScore = 18;
                } else if (avgTemp >= 55) {
                  tempScore = 8;
                }
                breakdown['Temperature'] = tempScore;
                totalScore += tempScore;

                // Cloud (Max 20)
                int cloudScore = 0;
                if (avgCloud <= 20) {
                  cloudScore = 20;
                } else if (avgCloud <= 40) {
                  cloudScore = 17;
                } else if (avgCloud <= 60) {
                  cloudScore = 12;
                } else if (avgCloud <= 80) {
                  cloudScore = 6;
                } else {
                  cloudScore = 2;
                }
                breakdown['Cloud Cover'] = cloudScore;
                totalScore += cloudScore;

                // Wind (Max 20)
                int windScore = 0;
                if (avgWind <= 5) {
                  windScore = 20;
                } else if (avgWind <= 10) {
                  windScore = 18;
                } else if (avgWind <= 15) {
                  windScore = 12;
                } else if (avgWind <= 20) {
                  windScore = 6;
                } else if (avgWind <= 24) {
                  windScore = 2;
                }
                breakdown['Wind Speed'] = windScore;
                totalScore += windScore;

                // Precip Prob (Max 15)
                int precipScore = 0;
                if (avgPrecipProb == 0) {
                  precipScore = 15;
                } else if (avgPrecipProb <= 10) {
                  precipScore = 12;
                } else if (avgPrecipProb <= 20) {
                  precipScore = 8;
                } else if (avgPrecipProb <= 35) {
                  precipScore = 4;
                } else if (avgPrecipProb <= 49) {
                  precipScore = 1;
                }
                breakdown['Precipitation'] = precipScore;
                totalScore += precipScore;

                // Humidity (Max 5)
                int humidityScore = (avgHumidity >= 30 && avgHumidity <= 70) ? 5 : 0;
                breakdown['Humidity'] = humidityScore;
                totalScore += humidityScore;

                // Time Bonus Removed from Score
                breakdown['Time Bonus'] = 0; // Keeping key to avoid UI crash if it expects it, or I should update UI. 
                // Better to remove key but updating UI is safer. I will update UI next.
                // Actually, I'll delete the key from breakdown here and handle it in UI.


                // Override total score if kill conditions exist, but keep breakdown
                // if (issues.isNotEmpty) {
                //     totalScore = 0;
                // }

                windows.add(InspectionWindow(
                  startTime: DateTime.parse(times[i]),
                  endTime: DateTime.parse(times[i]).add(Duration(hours: 2)), // 2 hour window
                  score: totalScore,
                  tempF: avgTemp,
                  windMph: avgWind,
                  cloudCover: avgCloud,
                  precipProb: avgPrecipProb,
                  humidity: avgHumidity,
                  condition: _getConditionCode(segmentCodes[0]),
                  issues: issues,
                  scoreBreakdown: breakdown,
                ));
            }
        }
    }
    return windows;
  }
  
  String _getConditionCode(int code) {
      // WMO codes
      if (code == 0) return 'Clear';
      if (code <= 3) return 'Partly Cloudy';
      if (code <= 48) return 'Foggy';
      if (code <= 67) return 'Rainy';
      if (code <= 77) return 'Snowy';
      if (code <= 82) return 'Rain Showers';
      return 'Stormy';
  }
}

File: services/WeatherService.ts
export interface InspectionWindow {
    startTime: Date;
    endTime: Date;
    score: number;
    tempF: number;
    windMph: number;
    cloudCover: number;
    precipProb: number;
    humidity: number;
    condition: string;
    issues: string[];
    scoreBreakdown: Record<string, number>;
    displayHour: number;
    displayDate: string;
    
    // V2 Scoring System
    scoreV2: number;
    classificationV2: 'Optimal' | 'Viable' | 'Inadvisable';
    issuesV2: string[];
    scoreBreakdownV2: Record<string, number>;
    pressureHpa: number;
    pressureTrend: number;
    pressureDelta3hr: number;
}

export class WeatherService {
    private static WEATHER_API_URL = 'https://api.open-meteo.com/v1/forecast';

    /**
     * Get logic coordinates from Zip Code
     */
    static async getCoordinates(zip: string, country: string = 'us'): Promise<{ lat: number; lng: number }> {
        // Zippopotam requires lowercase country codes
        const countryCode = country.toLowerCase();

        let searchZip = zip;

        // Normalize postal codes for Zippopotam (it supports outward codes for some countries)
        if (countryCode === 'gb') {
            // UK: Use outward code (first part before space) e.g., "SW1A 1AA" -> "SW1A"
            searchZip = zip.split(' ')[0].trim();
        } else if (countryCode === 'ca') {
            // Canada: Use FSA (first 3 chars) e.g., "K1A 0B1" -> "K1A"
            searchZip = zip.substring(0, 3);
        } else if (countryCode === 'nl') {
            // Netherlands: Use numeric part (first 4 chars) e.g., "1012 JS" -> "1012"
            searchZip = zip.substring(0, 4);
        }

        const response = await fetch(`https://api.zippopotam.us/${countryCode}/${searchZip}`);
        if (response.ok) {
            const data = await response.json() as any;
            const place = data.places[0];
            return {
                lat: parseFloat(place.latitude),
                lng: parseFloat(place.longitude),
            };
        } else {
            throw new Error(`Invalid Postal Code for ${country.toUpperCase()}`);
        }
    }

    /**
     * Determine the best Open-Meteo model based on country code.
     * GFS for Americas, ICON for Europe, ECMWF IFS for rest of world.
     */
    static getModelForCountry(countryCode?: string): string {
        const code = (countryCode || 'us').toLowerCase();

        // Americas → GFS (same model as NWS/AccuWeather)
        const americas = ['us', 'ca', 'mx', 'br', 'ar', 'cl', 'co', 'pe', 've', 'ec', 'bo', 'py', 'uy', 'gf', 'sr', 'gy', 'bz', 'gt', 'hn', 'sv', 'ni', 'cr', 'pa', 'cu', 'jm', 'ht', 'do', 'pr', 'tt', 'bb', 'bs', 'ag', 'dm', 'gd', 'kn', 'lc', 'vc'];
        if (americas.includes(code)) return 'gfs_seamless';

        // Europe → ICON (German DWD model, best for Europe)
        const europe = ['gb', 'de', 'fr', 'it', 'es', 'pt', 'nl', 'be', 'at', 'ch', 'pl', 'cz', 'sk', 'hu', 'ro', 'bg', 'hr', 'si', 'rs', 'ba', 'me', 'mk', 'al', 'gr', 'dk', 'no', 'se', 'fi', 'ee', 'lv', 'lt', 'ie', 'is', 'lu', 'mt', 'cy', 'ua', 'by', 'md'];
        if (europe.includes(code)) return 'icon_seamless';

        // Rest of world (Australia, Asia, Africa, etc.) → ECMWF IFS
        return 'ecmwf_ifs025';
    }

    /**
     * Fetch raw weather data (Includes pressure_msl for storm tracking)
     */
    static async getWeatherForecast(lat: number, lng: number, elevation?: number, countryCode?: string): Promise<any> {
        const model = this.getModelForCountry(countryCode);
        let url = `${this.WEATHER_API_URL}?latitude=${lat}&longitude=${lng}&hourly=temperature_2m,relative_humidity_2m,precipitation_probability,precipitation,weathercode,cloudcover,windspeed_10m,pressure_msl,surface_pressure&daily=sunrise,sunset&temperature_unit=fahrenheit&wind_speed_unit=mph&precipitation_unit=inch&timezone=auto&forecast_days=7&models=${model}`;

        if (elevation !== undefined) {
            url += `&elevation=${elevation}`;
        }
        const response = await fetch(url);
        if (response.ok) {
            return response.json();
        } else {
            throw new Error('Failed to load weather data');
        }
    }

    /**
     * Calculate 1-hour inspection windows for beekeeping
     */
    static calculateForecast(weatherData: any, isTBH: boolean = false, isMetric: boolean = false): InspectionWindow[] {
        const windows: InspectionWindow[] = [];
        const hourly = weatherData.hourly;
        const daily = weatherData.daily || {};
        const dailyTimes = (daily.time || []) as string[];
        const sunrises = (daily.sunrise || []) as string[];
        const sunsets = (daily.sunset || []) as string[];

        const times = hourly.time as string[];
        const temps = hourly.temperature_2m as number[];
        const humidities = hourly.relative_humidity_2m as number[];
        const precipProbs = hourly.precipitation_probability as number[];
        const precips = hourly.precipitation as number[];
        const codes = hourly.weathercode as number[];
        const clouds = hourly.cloudcover as number[];
        const winds = hourly.windspeed_10m as number[];
        const pressures = (hourly.pressure_msl || []) as number[];
        const surfacePressures = (hourly.surface_pressure || hourly.pressure_msl || []) as number[];

        // Group indices by Date (yyyy-MM-dd)
        const dayIndices: Record<string, number[]> = {};
        for (let i = 0; i < times.length; i++) {
            const dayKey = times[i].slice(0, 10);
            if (!dayIndices[dayKey]) dayIndices[dayKey] = [];
            dayIndices[dayKey].push(i);
        }

        const targetStartHours = [6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17];

        // Iterate over each day
        for (const dayKey in dayIndices) {
            const indices = dayIndices[dayKey];

            for (const startHour of targetStartHours) {
                // Find index for this specific hour
                let startIndex: number | undefined;
                for (const idx of indices) {
                    const hour = parseInt(times[idx].slice(11, 13));
                    if (hour === startHour) {
                        startIndex = idx;
                        break;
                    }
                }

                if (startIndex !== undefined) {
                    const i = startIndex;

                    // Use raw hourly values directly
                    const temp = temps[i];
                    const wind = winds[i];
                    const cloud = clouds[i];
                    const precipProb = precipProbs[i];
                    const precip = precips[i];
                    const code = codes[i];
                    const humidity = humidities[i];
                    const pressure = pressures[i] || 1013.25;

                    // Compute barometric pressure rate of change
                    const prevPressure = i > 0 ? (pressures[i - 1] || pressure) : pressure;
                    const pressureTrend = prevPressure - pressure; // positive means dropping

                    const issues: string[] = [];

                    // Localization Helpers
                    const tempStr = (t: number) => isMetric ? `${Math.round((t - 32) * 5 / 9)}°C` : `${Math.round(t)}°F`;

                    if (temp < 55) issues.push(`Too Cold (< ${tempStr(55)})`);

                    if (wind > 24) {
                        const speedStr = isMetric ? `${Math.round(24 * 1.60934)}km/h` : `24mph`;
                        issues.push(`Too Windy (> ${speedStr})`);
                    }

                    if (precipProb > 49) issues.push("Rain Likely (> 49%)");
                    if (precip > 0.02) issues.push("Raining");
                    if ([95, 96, 99].includes(code)) issues.push("Stormy Weather");

                    // TBH-specific: High heat issue
                    if (isTBH && temp > 92) issues.push(`Temperature > ${tempStr(92)} (comb slump risk)`);

                    let totalScore = 0;
                    const breakdown: Record<string, number> = {};

                    // 1. Temperature (Max 40)
                    let tempScore = 0;
                    if (temp >= 75) tempScore = 40;
                    else if (temp >= 70) tempScore = 37;
                    else if (temp >= 65) tempScore = 33;
                    else if (temp >= 60) tempScore = 27;
                    else if (temp >= 57) tempScore = 18;
                    else if (temp >= 55) tempScore = 8;

                    // TBH-specific: High heat penalty
                    if (isTBH && temp > 80) {
                        const degreesAbove80 = temp - 80;
                        const penalty = Math.floor(degreesAbove80 / 5) * 10;
                        tempScore = Math.max(0, tempScore - penalty);
                    }

                    breakdown['Temperature'] = tempScore;
                    totalScore += tempScore;

                    // 2. Cloud Cover (Max 20)
                    let cloudScore = 0;
                    if (cloud <= 20) cloudScore = 20;
                    else if (cloud <= 40) cloudScore = 17;
                    else if (cloud <= 60) cloudScore = 12;
                    else if (cloud <= 80) cloudScore = 6;
                    else cloudScore = 6;
                    breakdown['Cloud Cover'] = cloudScore;
                    totalScore += cloudScore;

                    // 3. Wind (Max 20)
                    let windScore = 0;
                    if (wind <= 5) windScore = 20;
                    else if (wind <= 10) windScore = 18;
                    else if (wind <= 15) windScore = 12;
                    else if (wind <= 20) windScore = 6;
                    else if (wind <= 24) windScore = 2;
                    breakdown['Wind Speed'] = windScore;
                    totalScore += windScore;

                    // 4. Precipitation Probability (Max 15)
                    let precipScore = 0;
                    if (precipProb === 0) precipScore = 15;
                    else if (precipProb <= 10) precipScore = 12;
                    else if (precipProb <= 20) precipScore = 8;
                    else if (precipProb <= 35) precipScore = 4;
                    else if (precipProb <= 49) precipScore = 1;
                    breakdown['Precipitation'] = precipScore;
                    totalScore += precipScore;

                    // 5. Humidity (Max 5)
                    const humidityScore = (humidity >= 30 && humidity <= 70) ? 5 : 0;
                    breakdown['Humidity'] = humidityScore;
                    totalScore += humidityScore;

                    // ============================================
                    // ============================================
                    // V2 DECISION MATRIX CALCULATION (0-9 POINTS)
                    // ============================================
                    const issuesV2: string[] = [];

                    // Dynamic Sunrise/Sunset & Wake-up Calculations
                    const dailyIdx = dailyTimes.indexOf(dayKey);
                    let sunsetHour = 18; // Fallback to 6:00 PM
                    let sunsetMinute = 0;
                    if (dailyIdx !== -1 && sunsets[dailyIdx]) {
                        const sunsetStr = sunsets[dailyIdx];
                        sunsetHour = parseInt(sunsetStr.slice(11, 13));
                        sunsetMinute = parseInt(sunsetStr.slice(14, 16));
                    }

                    const slotEndHour = startHour + 1;
                    const sunsetTimeInMinutes = sunsetHour * 60 + sunsetMinute;
                    const slotEndTimeInMinutes = slotEndHour * 60;
                    const isSafeBeforeSunset = slotEndTimeInMinutes <= (sunsetTimeInMinutes - 60);

                    const temp1HourAgo = i > 0 ? (temps[i - 1] ?? temp) : temp;
                    const isWarmEnough1HourAgo = temp1HourAgo >= 60;

                    // Step 1: Check Fail-Safes (Short-Circuit Hard Aborts)
                    if (temp < 57) {
                        issuesV2.push(`Brood Chill Threshold Triggered (Temp < 57°F / 14°C)`);
                    }
                    if (temp > 92) {
                        issuesV2.push(`Comb Heat/Heat Stroke Threshold Triggered (Temp > 92°F / 33°C)`);
                    }
                    
                    const hasRainV2 = precip > 0.02 || [95, 96, 99].includes(code) || (code >= 51 && code <= 67) || (code >= 80 && code <= 82) || precipProb >= 50;
                    if (hasRainV2) {
                        issuesV2.push(`Active Precipitation Triggered`);
                    }

                    if (wind > 18) {
                        const wStr = isMetric ? `${Math.round(18 * 1.60934)}km/h` : `18mph`;
                        issuesV2.push(`Flight Disruption Wind Triggered (Wind > ${wStr})`);
                    }

                    if (!isWarmEnough1HourAgo) {
                        issuesV2.push(`Bees not awake (temperature 1 hour ago was ${Math.round(temp1HourAgo)}°F, must be 60°F+ for at least 1 hour)`);
                    }

                    if (!isSafeBeforeSunset) {
                        const sunsetTimeStr = `${sunsetHour > 12 ? sunsetHour - 12 : sunsetHour}:${sunsetMinute.toString().padStart(2, '0')}${sunsetHour >= 12 ? 'pm' : 'am'}`;
                        issuesV2.push(`Too close to sunset (must end at least 1 hour before sunset at ${sunsetTimeStr} to allow foragers to return)`);
                    }

                    // Step 2: High-Desert Barometric Pressure Velocity Calculation
                    // Helper to compute a 3-hour trailing moving average smoothed station pressure
                    const getSmoothedSurfacePressure = (index: number): number => {
                        const p1 = surfacePressures[index] || pressure;
                        const p2 = index > 0 ? (surfacePressures[index - 1] || p1) : p1;
                        const p3 = index > 1 ? (surfacePressures[index - 2] || p2) : p2;
                        return (p1 + p2 + p3) / 3;
                    };

                    const s_current = getSmoothedSurfacePressure(i);
                    const s_oldest = i >= 3 ? getSmoothedSurfacePressure(i - 3) : getSmoothedSurfacePressure(0);
                    const pressureDelta3hr = s_oldest - s_current; // positive means dropping
                    
                    let pressure_penalty = 0;
                    if (pressureDelta3hr >= 4.0) {
                        issuesV2.push(`Imminent severe barometric front (Delta P = ${pressureDelta3hr.toFixed(1)} mb)`);
                    } else if (pressureDelta3hr >= 1.5) {
                        pressure_penalty = -2;
                    }

                    // Step 3: Points calculations & Weighted Scoring Matrix
                    let tempPts = 0;
                    let timePts = 0;
                    let skyPts = 0;
                    let windPts = 0;

                    // Temperature (Optimal: 68-85 -> +3; Sub-optimal: 58-67 or 86-91 -> +1)
                    if (temp >= 68 && temp <= 85) tempPts = 3;
                    else if ((temp >= 58 && temp <= 67) || (temp >= 86 && temp <= 91)) tempPts = 1;
                    
                    // Time of Day (Optimal: >= 60°F 1 hour ago AND >= 1 hour before sunset -> +2; Sub-optimal: >= 55°F 1 hour ago -> +1; else 0)
                    if (isSafeBeforeSunset) {
                        if (temp1HourAgo >= 60) timePts = 2;
                        else if (temp1HourAgo >= 55) timePts = 1;
                    }
                    
                    // Wind (Optimal: <10mph -> +2; Sub-optimal: 10-15mph -> +1; else 0)
                    if (wind < 10) windPts = 2;
                    else if (wind >= 10 && wind <= 15) windPts = 1;

                    // Sky Condition (Clear: <30% -> +2; Partly Cloudy: 30-70% -> +1; else 0)
                    if (cloud < 30) skyPts = 2;
                    else if (cloud >= 30 && cloud <= 70) skyPts = 1;

                    // Apply Barometric Pressure Penalty
                    const scoreV2 = Math.max(0, tempPts + timePts + skyPts + windPts + pressure_penalty);
                    
                    // Determine Suitability Classification
                    let classificationV2: 'Optimal' | 'Viable' | 'Inadvisable' = 'Inadvisable';
                    if (issuesV2.length > 0) {
                        // Hard Abort triggered: classification is forced to Inadvisable to show Red Cell warning,
                        // but points and scores are fully calculated and displayed!
                        classificationV2 = 'Inadvisable';
                    } else {
                        if (scoreV2 >= 7) classificationV2 = 'Optimal';
                        else if (scoreV2 >= 4) classificationV2 = 'Viable';
                        else classificationV2 = 'Inadvisable';
                    }

                    const breakdownV2 = {
                        'Temperature': tempPts,
                        'Time of Day': timePts,
                        'Sky Condition': skyPts,
                        'Wind Speed': windPts
                    };

                    windows.push({
                        startTime: new Date(times[i]),
                        endTime: new Date(new Date(times[i]).getTime() + 1 * 60 * 60 * 1000), // +1 hour
                        score: totalScore,
                        tempF: temp,
                        windMph: wind,
                        cloudCover: cloud,
                        precipProb: precipProb,
                        humidity: humidity,
                        condition: this.getConditionCode(code),
                        issues: issues,
                        scoreBreakdown: breakdown,
                        displayHour: startHour,
                        displayDate: dayKey,
                        
                        // New V2 decision matrix output
                        scoreV2,
                        classificationV2,
                        issuesV2,
                        scoreBreakdownV2: breakdownV2,
                        pressureHpa: pressure,
                        pressureTrend,
                        pressureDelta3hr
                    });
                }
            }
        }
        return windows;
    }

    static getConditionCode(code: number): string {
        // WMO codes
        if (code === 0) return 'Clear';
        if (code <= 3) return 'Partly Cloudy';
        if (code <= 48) return 'Foggy';
        if (code <= 67) return 'Rainy';
        if (code <= 77) return 'Snowy';
        if (code <= 82) return 'Rain Showers';
        return 'Stormy';
    }
}
